op-e2e/actions/l2_verifier.go

```diff
@@ -88,82 +88,98 @@ func NewL2Verifier(t Testing, log log.Logger, l1 derive.L1Fetcher, blobsSrc deri
 	ctx, cancel := context.WithCancel(context.Background())
 	t.Cleanup(cancel)
 
-	rootDeriver := &event.DeriverMux{}
-	var synchronousEvents event.EmitterDrainer
-	synchronousEvents = event.NewQueue(log, ctx, rootDeriver, event.NoopMetrics{})
-	synchronousEvents = event.NewLimiterDrainer(ctx, synchronousEvents, rate.Limit(1000), 20, func() {
-		log.Warn("Hitting events rate-limit. An events code-path may be hot-looping.")
-		t.Fatal("Tests must not hot-loop events")
-	})
+	executor := event.NewGlobalSynchronous()
+	sys := event.NewSystem(log, executor)
+	t.Cleanup(sys.Stop)
+	opts := event.DefaultRegisterOpts()
+	opts.Emitter = event.EmitterOpts{
+		Limiting: true,
+		Rate:     rate.Limit(1000),
+		Burst:    1000,
+		OnLimited: func() {
+			log.Warn("Hitting events rate-limit. An events code-path may be hot-looping.")
+			t.Fatal("Tests must not hot-loop events")
+		},
+	}
 
 	metrics := &testutils.TestDerivationMetrics{}
-	ec := engine.NewEngineController(eng, log, metrics, cfg, syncCfg, synchronousEvents)
-	engineResetDeriver := engine.NewEngineResetDeriver(ctx, log, cfg, l1, eng, syncCfg, synchronousEvents)
+	var ec *engine.EngineController
+	event.Setup(sys, "engine-controller", opts, func(em event.Emitter) event.Deriver {
+		ec = engine.NewEngineController(eng, log, metrics, cfg, syncCfg, em)
+		return nil // does not respond to events yet
+	})
+	event.Setup(sys, "engine-reset", opts, func(em event.Emitter) event.Deriver {
+		return engine.NewEngineResetDeriver(ctx, log, cfg, l1, eng, syncCfg, em)
+	})
 
-	clSync := clsync.NewCLSync(log, cfg, metrics, synchronousEvents)
+	var clSync driver.CLSync
+	event.Setup(sys, "cl-sync", opts, func(em event.Emitter) event.Deriver {
```

Comment:
I have been thinking about adding an optional `.SetEmitter()` interface check on the `Deriver` argument, and to remove `Setup()`, to mutate the object rather than do this defer trick. Mutation is not as elegant, but does prevent this function nesting and defer thing, which may just be worth it.

Reply:
Added a commit that implements the `AttachEmitter(em event.Emitter)` check upon registration on the deriver, so the setup is cleaner and can happen with the one and only `system.Register(...)`.
If that looks good to you then we can resolve this + put on merge train.